documentation/pwa/tutorial-pwa-offline.asciidoc

```diff
@@ -1,23 +1,88 @@
 ---
-title: PWA Offline Page
+title: PWA Offline Experience
 order: 6
 layout: page
 ---
 
-= PWA Offline Page
+= PWA Offline Experience
 
-Vaadin automatically generates and serves an offline page. This is a simple page that:
+Vaadin supports two alternative ways of building offline experience:
 
-* Includes the application name and icon.
-* Communicates to the user that the application is offline, because there is no network connection.
+* Client-side TypeScript views (default)
+* A separate offline page
+
+For PWAs built with Vaadin, the service worker provides offline support for TypeScript routes and views. By default, it stores the application shell HTML, the compiled frontend bundles, and the other necessary resources, and then serves them offline from the browser’s cache.
+
+You can optionally use a separate offline page instead of TypeScript views offline (`offlinePath` property in `@PWA` annotation).
+
+== Offline TypeScript Views
+
+Adding the <<tutorial-pwa-pwa-with-flow#,`@PWA` annotation>> on your application shell class enables the service worker, which will automatically serve the client-side views offline.
+
+The service worker will also cache and serve offline all the imported dependencies (using `import`) in TypeScript views.
+
+[WARNING]
+====
+Network access is not available offline.
+
+For TypeScript views using endpoints, working offline requires fallback responses:
+
+[source,typescript]
+----
+import {EndpointError} from "@vaadin/flow-frontend";
+
+// import the remote endpoint
+import * as dataEndpoint from '../../generated/DataEndpoint';
+
+// Wrap endpoint calls to return fallback data when offline
+async function getViewData() {
+  try {
+    return await dataEndpoint.getViewData();
+  } catch (e) {
+    if (!(e instanceof EndpointError)) {
+      // network failure, return fallback data
+      return [];
+    }
+  }
+}
+----
+====
 
 == Creating a Custom Offline Page
 
-To override the default offline page:
+To use a separate offline page:
 
 . Create a file named `offline.html`.
-. Add the file to your `src/main/webapp/` folder. 
+. Add the file to your `src/main/webapp/` folder.
```

Comment:
better to change this folder to another one recommended by spring? see [this comment](https://github.com/vaadin/flow/issues/8996#issuecomment-705576552)